### PR TITLE
Added Version Number Document

### PR DESCRIPTION
**Pull request by VenkyOS**
Signed-off-by: Venkataraman Padmanabhan <venkataraman.padmanabhan@seagate.com>

### CODE REVIEW

File: doc/HLD_Version_Numbers.rst [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+.. image:: Images/scev2.PNG
+
+Scenario 3
+
+.. image:: Images/scev3.PNG
+
+Scenario 4
+
+.. image:: Images/scev4.PNG
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:216]: Images are missing.

Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst:216]: I have added them now.

File: doc/HLD_Version_Numbers.rst [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+***************
+
+See version number comparison invariant above. 
+
+Use Cases (Scenarios)
+=====================
+
+Scenario 1
+
+.. image:: Images/scev1.PNG
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:204]: Remove all the IMAGES here, and replace them with tables.

File: doc/HLD_Version_Numbers.rst [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+This specification is all about handling failures. 
+
+Analysis
+===============
+
+The hybrid operation mode introduced above, disentangles issues of resource management and concurrency control from failure recovery, which traditional session-slot based protocols fail to do. It's also instructive to compare version maintenance with the epoch algorithm: version numbers are, roughly speaking, per-object epochs.
+
+
+
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:231]: The paragraphs and titles, are organized in a properly level.
They are all correct.
Please follow this one to modify PR #193, PR #196 and PR #201

**Review by mukul-seagate11 — APPROVED** [review]
OK

**Comment by johnbent** [comment]
This looks great to me!  Nice job!

File: doc/HLD_Version_Numbers.rst [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+***************
+Definitions
+*************** 
+
+See the Glossary for general M0 definitions and HLD of FOL for the definitions of file system operation, update and lsn. The following additional definitions are required:
+
+- for the purposes of the present design it is assumed that a file system update acts on units1. For example, a typical meta-data update acts on one or more "inodes" and a typical data update acts on inodes and data blocks. Inodes, data blocks, directory entries, etc. are all examples of units. It is further assumed that units involved in an update are unambiguously identified2 and that complete file system state is a disjoint union of states of comprising units. (Of course, there are consistency relationships between units, e.g., inode nlink counter must be consistent with contents of directories in the name-space).
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:21]: What is the last digit here in "update acts on units1"?
I think this is a reference link?
If yes, it should be like this:
```
update acts on units[1]
```

There are multiple places with this kind of error.

File: doc/HLD_Version_Numbers.rst [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+- [a version number in an update stream] cache re-integration uses version numbers to implement update streams. For each update stream between a client node C and a server node S, a special stream unit U is created. This unit is persistently stored on S and cached on C. Each update cached on C and targeted for re-integration on S modifies U. Therefore, each update sent from C to S contains U's version counter which is incremented by each update. This counter acts like a xid of the traditional Lustre protocol, and (U, U.VC) pair acts as a (session, slot) pair of the NFSv4 session protocol. The collection of these special units for all update streams is similar to last_rcvd file of Lustre and to EOS cache of NFSv4.
+
+*********************
+Logical Specification
+*********************
+
+Ordering
+=========
+
+Internally, version number is defined as a two-component data-structure (struct M0_verno), with little internal state or logic.
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:86]: s/struct M0_verno/struct m0_verno/

File: doc/HLD_Version_Numbers.rst (hunk 2 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 3 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst:328]: Sure

File: doc/HLD_Version_Numbers.rst (hunk 4 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 5 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst]: Incorporated.

File: doc/HLD_Version_Numbers.rst [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+.. image:: Images/rev1.PNG
+
+becomes after open("foo") call: 
+
+.. image:: Images/rev2.PNG
+
+- [a version number applicability check] distributed transaction manager uses version numbers for recovery: when an update is replayed on a node, a modification to an unit should be replayed if and only if before-version shipped with an update is the same as before-version stored in the unit;
+
+- [a version number in an update stream] cache re-integration uses version numbers to implement update streams. For each update stream between a client node C and a server node S, a special stream unit U is created. This unit is persistently stored on S and cached on C. Each update cached on C and targeted for re-integration on S modifies U. Therefore, each update sent from C to S contains U's version counter which is incremented by each update. This counter acts like a xid of the traditional Lustre protocol, and (U, U.VC) pair acts as a (session, slot) pair of the NFSv [4] session protocol. The collection of these special units for all update streams is similar to last_rcvd file of Lustre and to EOS cache of NFSv [4].
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:77]: NFSv4 is a single word.

File: doc/HLD_Version_Numbers.rst [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+- A unit version number is an additional piece of information attached to the unit. A version number is drawn from some linearly ordered domain. A version number changes on every update of the unit state in such a way that ordering of unit states in the serial history can be deduced by comparing version numbers associated with the corresponding states.    
+
+***************
+Requirements
+***************
+
+- [r.verno.serial]: version numbers order unit states in the unit serial history; 
+
+- [r.verno.resource]: version numbers are managed as a scalable distributed resource [3]: entities caching a given unit could also cache4 some range of version numbers and to generate version numbers for updates of local cached state locally;
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:33]: entities caching a given unit could also cache4
->
entities caching a given unit could also cache[4]

File: doc/HLD_Version_Numbers.rst [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+
+******************
+Design Highlights
+******************
+
+In the presence of caching, requirements [r.verno.resource] and [r.verno.fol] are seemingly contradictory: if two caching client nodes assigned (as allowed by [r.verno.resource]) version numbers to two independent units, then after re-integration of units to their common master server, the version numbers must refer to the master's fol, but clients cannot produce such references without extremely inefficient serialization of all accesses to the units on the server. 
+
+To deal with that, a version number is made compound: it consists of two components: 
+
+- LSN7: a reference to a fol record, corresponding to the unit state;
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:51]: LSN7
->
LSN [7]

File: doc/HLD_Version_Numbers.rst [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+*********************
+
+Ordering
+=========
+
+Internally, version number is defined as a two-component data-structure (struct m0_verno), with little internal state or logic. 
+
+The following invariant, referred to as a version number comparison invariant is maintained:
+
+where v0 and v1 are two version numbers for the same unit (taken on the same node), and lsn comparison function is defined by the FOL HLD11. This invariant means that per-object and per-fol (i.e., per-node) orderings of updates are compatible.
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:90]: the FOL HLD11
->
the FOL HLD[11]

File: doc/HLD_Version_Numbers.rst [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+***************
+
+This specification is all about handling failures. 
+
+Analysis
+===============
+
+The hybrid operation mode introduced above, disentangles issues of resource management and concurrency control from failure recovery, which traditional session-slot based protocols fail to do. It's also instructive to compare version maintenance with the epoch algorithm: version numbers are, roughly speaking, per-object epochs.
+
+
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:327]: The references should be listed here.
But they are missing.

File: doc/HLD_Version_Numbers.rst [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+***************
+Definitions
+*************** 
+
+See the Glossary for general M0 definitions and HLD of FOL for the definitions of file system operation, update and lsn. The following additional definitions are required:
+
+- for the purposes of the present design it is assumed that a file system update acts on units [1]. For example, a typical meta-data update acts on one or more "inodes" and a typical data update acts on inodes and data blocks. Inodes, data blocks, directory entries, etc. are all examples of units. It is further assumed that units involved in an update are unambiguously identified [2] and that complete file system state is a disjoint union of states of comprising units. (Of course, there are consistency relationships between units, e.g., inode nlink counter must be consistent with contents of directories in the name-space).
```
Comment (huanghua78) [inline doc/HLD_Version_Numbers.rst:21]: >- for the purposes of the present design it is assumed that a file system update acts on units [1].

I am sorry, in the last review comments, I said this should be changed to [1].
But this was wrong.
When I looked at the original doc, this should be a link to some internal requirement.
Please check with the original doc to correct all the internal links.

And keep the [0] and [1] of internal links listed at the end of this document.

Comment (huanghua78) [reply doc/HLD_Version_Numbers.rst:21]: Please follow and compare with the original document to correct these.

It is not a copy & paste to convert a doc.

File: doc/HLD_Version_Numbers.rst (hunk 6 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 7 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst:33]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 8 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst:51]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 9 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 11 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst:21]: Incorporated.

File: doc/HLD_Version_Numbers.rst (hunk 10 above)
Comment (VenkyOS) [reply doc/HLD_Version_Numbers.rst:334]: Incorporated.

**Review by huanghua78 — APPROVED** [review]

**Pull request merged**